firedrake/adjoint/projection.py

```diff
@@ -0,0 +1,34 @@
+from pyadjoint.tape import annotate_tape, stop_annotating, get_working_tape
+from pyadjoint.overloaded_type import create_overloaded_object
+from firedrake.adjoint.blocks import SolveBlock, ProjectBlock
+
+
+def annotate_project(project):
+
+    def wrapper(*args, **kwargs):
+        """The project call performs an equation solve, and so it too must be annotated so that the
+        adjoint and tangent linear models may be constructed automatically by pyadjoint.
+
+        To disable the annotation of this function, just pass :py:data:`annotate=False`. This is useful in
+        cases where the solve is known to be irrelevant or diagnostic for the purposes of the adjoint
+        computation (such as projecting fields to other function spaces for the purposes of
+        visualisation)."""
+
+        annotate = annotate_tape(kwargs)
+        with stop_annotating():
+            output = project(*args, **kwargs)
+        output = create_overloaded_object(output)
```

Comment:
Non-annotated outputs should not be overloaded, I think.

Reply:
No, we overload in all cases. The alternative is very unsafe.